pallets/automation-time/src/fees.rs

```diff
@@ -0,0 +1,90 @@
+// This file is part of OAK Blockchain.
```

Comment:
Too many changes for github to recognize `exchange.rs -> fees.rs`. Including relevant diff to help with reviewing.
```diff
diff --git a/pallets/automation-time/src/exchange.rs b/pallets/automation-time/src/exchange.rs
index 95c730b..b569c79 100644
--- a/pallets/automation-time/src/exchange.rs
+++ b/pallets/automation-time/src/exchange.rs
@@ -16,63 +16,28 @@
 // limitations under the License.
 
 /// ! Traits and default implementation for paying execution fees.
-use crate::Config;
+use crate::{BalanceOf, Config};
 
-use codec::FullCodec;
 use sp_runtime::{
-	traits::{AtLeast32BitUnsigned, CheckedSub, MaybeSerializeDeserialize, Zero},
+	traits::{CheckedSub, Zero},
 	DispatchError,
 	TokenError::BelowMinimum,
 };
-use sp_std::{fmt::Debug, marker::PhantomData};
+use sp_std::marker::PhantomData;
 
-use frame_support::traits::{
-	Currency, ExistenceRequirement, Imbalance, OnUnbalanced, WithdrawReasons,
-};
+use frame_support::traits::{Currency, ExistenceRequirement, OnUnbalanced, WithdrawReasons};
 
-type NegativeImbalanceOf<C, T> =
-	<C as Currency<<T as frame_system::Config>::AccountId>>::NegativeImbalance;
+type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
 
 /// Handle withdrawing, refunding and depositing of transaction fees.
-pub trait NativeTokenExchange<T: Config> {
-	/// The underlying integer type in which fees are calculated.
-	type Balance: AtLeast32BitUnsigned
-		+ FullCodec
-		+ Copy
-		+ MaybeSerializeDeserialize
-		+ Debug
-		+ Default
-		+ scale_info::TypeInfo;
-	type LiquidityInfo: Default;
-
-	/// The minimum balance any single account may have. This is equivalent to the `Balances`
-	/// module's `ExistentialDeposit`.
-	fn minimum_balance() -> Self::Balance;
-
-	///Transfer some liquid free balance to another staker.
-	/// This is a very high-level function. It will ensure all appropriate fees are paid and no imbalance in the system remains.
-	fn transfer(
-		source: &T::AccountId,
-		dest: &T::AccountId,
-		value: Self::Balance,
-	) -> Result<(), DispatchError>;
-
-	/// Deposit some `value` into the free balance of `who`, possibly creating a new account.
-	///
-	/// This function is a no-op if:
-	/// - the `value` to be deposited is zero; or
-	/// - the `value` to be deposited is less than the required ED and the account does not yet
-	///   exist; or
-	/// - the deposit would necessitate the account to exist and there are no provider references;
-	///   or
-	/// - `value` is so large it would cause the balance of `who` to overflow.
-	fn deposit_creating(who: &T::AccountId, value: Self::Balance);
-
+pub trait HandleFees<T: Config> {
 	/// Ensure the fee can be paid.
-	fn can_pay_fee(who: &T::AccountId, fee: Self::Balance) -> Result<(), DispatchError>;
+	fn can_pay_fee(who: &T::AccountId, fee: BalanceOf<T>) -> Result<(), DispatchError>;
 
 	/// Once the task has been scheduled we need to charge for the execution cost.
-	fn withdraw_fee(who: &T::AccountId, fee: Self::Balance) -> Result<(), DispatchError>;
+	fn withdraw_fee(who: &T::AccountId, fee: BalanceOf<T>) -> Result<(), DispatchError>;
 }
 
 /// Implements the transaction payment for a pallet implementing the `Currency`
@@ -81,74 +46,40 @@ pub trait NativeTokenExchange<T: Config> {
 ///
 /// The unbalance handler is given 2 unbalanceds in [`OnUnbalanced::on_unbalanceds`]: fee and
 /// then tip.
-pub struct CurrencyAdapter<C, OU>(PhantomData<(C, OU)>);
+pub struct FeeHandler<OU>(PhantomData<OU>);
 
-/// Default implementation for a Currency and an OnUnbalanced handler.
+/// Default implementation for an OnUnbalanced handler.
 ///
 /// The unbalance handler is given 2 unbalanceds in [`OnUnbalanced::on_unbalanceds`]: fee and
 /// then tip.
-impl<T, C, OU> NativeTokenExchange<T> for CurrencyAdapter<C, OU>
+impl<T, OU> HandleFees<T> for FeeHandler<OU>
 where
 	T: Config,
-	C: Currency<<T as frame_system::Config>::AccountId>,
-	C::PositiveImbalance: Imbalance<
-		<C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
-		Opposite = C::NegativeImbalance,
-	>,
-	C::NegativeImbalance: Imbalance<
-		<C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
-		Opposite = C::PositiveImbalance,
-	>,
-	OU: OnUnbalanced<NegativeImbalanceOf<C, T>>,
+	OU: OnUnbalanced<NegativeImbalanceOf<T>>,
 {
-	type LiquidityInfo = Option<NegativeImbalanceOf<C, T>>;
-	type Balance = <C as Currency<<T as frame_system::Config>::AccountId>>::Balance;
-
-	/// The minimum balance any single account may have. This is equivalent to the `Balances`
-	/// module's `ExistentialDeposit`.
-	fn minimum_balance() -> Self::Balance {
-		C::minimum_balance()
-	}
-
-	///Transfer some liquid free balance to another staker.
-	/// This is a very high-level function. It will ensure all appropriate fees are paid and no imbalance in the system remains.
-	fn transfer(
-		source: &T::AccountId,
-		dest: &T::AccountId,
-		value: Self::Balance,
-	) -> Result<(), DispatchError> {
-		C::transfer(source, dest, value, ExistenceRequirement::KeepAlive)?;
-		Ok(())
-	}
-
-	// Creates new account and deposits balance into account
-	fn deposit_creating(who: &T::AccountId, value: Self::Balance) {
-		C::deposit_creating(who, value);
-	}
-
 	// Ensure the fee can be paid.
-	fn can_pay_fee(who: &T::AccountId, fee: Self::Balance) -> Result<(), DispatchError> {
+	fn can_pay_fee(who: &T::AccountId, fee: BalanceOf<T>) -> Result<(), DispatchError> {
 		if fee.is_zero() {
 			return Ok(())
 		}
 
-		let free_balance = C::free_balance(who);
+		let free_balance = T::Currency::free_balance(who);
 		let new_amount =
 			free_balance.checked_sub(&fee).ok_or(DispatchError::Token(BelowMinimum))?;
-		C::ensure_can_withdraw(who, fee, WithdrawReasons::FEE, new_amount)?;
+		T::Currency::ensure_can_withdraw(who, fee, WithdrawReasons::FEE, new_amount)?;
 
 		Ok(())
 	}
 
 	/// Withdraw the fee.
-	fn withdraw_fee(who: &T::AccountId, fee: Self::Balance) -> Result<(), DispatchError> {
+	fn withdraw_fee(who: &T::AccountId, fee: BalanceOf<T>) -> Result<(), DispatchError> {
 		if fee.is_zero() {
 			return Ok(())
 		}
 
 		let withdraw_reason = WithdrawReasons::FEE;
 
-		match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+		match T::Currency::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
 			Ok(imbalance) => {
 				OU::on_unbalanceds(Some(imbalance).into_iter());
 				Ok(())
```